### PR TITLE
Use merge pattern for Archival queries

### LINKED ISSUE #938: Standardize timestamp column types in the Archival materialization
### Issue description
Running `dbt archive` (on Snowflake) throws a cryptic error related to timezone types. 

### Results
```
16:38:45 | Concurrency: 2 threads (target='dev')
16:38:45 |
16:38:45 | 1 of 1 START archive dbt_adam.locations --> dbt_adam_archive.locations_archived [RUN]
16:38:56 | 1 of 1 ERROR archiving dbt_adam.locations --> dbt_adam_archive.locations_archived [ERROR in 11.08s]
16:38:56 |
16:38:56 | Finished running 1 archives in 14.58s.

Completed with 1 errors:

Database Error in archive locations_archived (dbt_project.yml)
  001790 (42601): 1d165133-4c2a-4587-98a4-995034f1a2e0: SQL compilation error:
  inconsistent data type for result columns for set operator input branches, expected TIMESTAMP_LTZ(9), got TIMESTAMP_NTZ(9) for expression [{2}] branch {3}
```

### System information
The output of `dbt --version`:
```
installed version: 0.10.2
   latest version: 0.10.2

Up to date!
```

The operating system you're running on: OS X High Sierra 10.13.6

The python version you're using: Python 3.5.4

### Steps to reproduce
Slack discussion here: https://getdbt.slack.com/archives/C2JRRQDTL/p1534508532000100

When I first got the error, I then cast my `created_at` field in the source table as timestamp_ltz, and verified that Snowflake returned that data type when querying the source table. It did not fix the error, though. It could be something related to the `null::timestamp` command. This is in the last query that failed; however, changing `null:timestamp` to `null:timestamp_ltz`, etc, doesn't fix the error. 

This is the query that failed: 
```
2018-08-17 16:38:55,632 (Thread-1): On locations_archived: create temporary table
    locations_archived__dbt_archival_tmp
  as (
    with dbt_archive_sbq as (
        

    

    with current_data as (

        select
            
                "ID" ,
            
                "UUID" ,
            
                "PERSON_ID" ,
            
                "NAME" ,
            
                "ADDRESS_ID" ,
            
                "CREATED_AT" ,
            
                "UPDATED_AT" ,
            
                "PHONE_NUMBER" ,
            
                "USED_FOR" ,
            
                "LANGUAGE_PREFERENCE" ,
            
                "DEVICE_TYPE" ,
            
                "DEFAULT_LANG_PREF" 
            ,
            updated_at as "dbt_updated_at",
            id as "dbt_pk",
            updated_at as "valid_from",
            null::TIMESTAMP as "tmp_valid_to"
        from "DBT_ADAM"."LOCATIONS"

    ),

    archived_data as (

        select
            
                "ID",
            
                "UUID",
            
                "PERSON_ID",
            
                "NAME",
            
                "ADDRESS_ID",
            
                "CREATED_AT",
            
                "UPDATED_AT",
            
                "PHONE_NUMBER",
            
                "USED_FOR",
            
                "LANGUAGE_PREFERENCE",
            
                "DEVICE_TYPE",
            
                "DEFAULT_LANG_PREF",
            
            updated_at as "dbt_updated_at",
            id as "dbt_pk",
            "valid_from",
            "valid_to" as "tmp_valid_to"
        from "DBT_ADAM_ARCHIVE"."LOCATIONS_ARCHIVED"

    ),

    insertions as (

        select
            current_data.*,
            null::TIMESTAMP as "valid_to"
        from current_data
        left outer join archived_data
          on archived_data."dbt_pk" = current_data."dbt_pk"
        where archived_data."dbt_pk" is null or (
          archived_data."dbt_pk" is not null and
          current_data."dbt_updated_at" > archived_data."dbt_updated_at" and
          archived_data."tmp_valid_to" is null
        )
    ),

    updates as (

        select
            archived_data.*,
            current_data."dbt_updated_at" as "valid_to"
        from current_data
        left outer join archived_data
          on archived_data."dbt_pk" = current_data."dbt_pk"
        where archived_data."dbt_pk" is not null
          and archived_data."dbt_updated_at" < current_data."dbt_updated_at"
          and archived_data."tmp_valid_to" is null
    ),

    merged as (

      select *, 'update' as "change_type" from updates
      union all
      select *, 'insert' as "change_type" from insertions

    )

    select *,
        
  
    md5("dbt_pk" || '|' || "dbt_updated_at")

 as "scd_id"
    from merged


      )
      select * from dbt_archive_sbq
  );
2018-08-17 16:38:56,228 (Thread-1): Snowflake error: 001790 (42601): 1d165133-4c2a-4587-98a4-995034f1a2e0: SQL compilation error:
inconsistent data type for result columns for set operator input branches, expected TIMESTAMP_LTZ(9), got TIMESTAMP_NTZ(9) for expression [{2}] branch {3}
```

### LINKED ISSUE #1339: Implement archival with a "merge" statement
### Feature description

The current implementation of archival (0.13.0) effectively implements a `merge` using `update` and `insert` statements. Instead, we should leverage a `merge` abstraction (also used by incremental models) to help normalize the implementation of archival.

There are a few benefits of using a merge here:
1. it is an atomic way of performing archival on databases like Snowflake[1] and BigQuery
2. the database can presumably do less work if the inserts and updates are specified in the same query
3. It should greatly simplify the archival materialization sql

Reasons not to do this:
1. if it ain't broke.....
2. `merge` is not implemented for all adapters. We'll need to build out a `merge` abstraction for redshift/postgres/et al, which could be complicated

### [1] Original issue description, snowflake specific
If two archive jobs run simultaneously on Snowflake, duplicate records can be inserted into the archive destination table. This problem can be circumvented with a merge.

The problem here is that archival is currently implemented as:
1. create temp table
2. insert
3. update

If two jobs run at the same time, they will both create identical temp tables (how do these not conflict?). When the jobs proceed to insert/update data, they will both duplicate work in the insert + update steps, resulting in duplicated data being inserted into the destination table.

Because a proper Snowflake merge would happen as a single atomic operation, two merges that are serialized would still result in the intended behavior. In this approach, dbt wouldn’t use a temp table. Instead, the merge would be responsible for finding new records to merge, inserting, and updating all at once. The second serialized merge would find no changes to merge, and would exit without modifying the destination table.

### LINKED ISSUE #1423: DBT Archive Failing in BiqQuery
## Issue

Currently the DBT archive feature uses [temporary archive](https://github.com/fishtown-analytics/dbt/blob/dev/wilt-chamberlain/plugins/bigquery/dbt/adapters/bigquery/impl.py#L337-L351) features in bigquery. When tables are "large enough" then this fails due to [Maximum response size](https://cloud.google.com/bigquery/quotas#queries). 

If we can write to a table in BQ instead of a temporary table, then delete it this may solve the issue, as `The maximum response size is unlimited when writing large query results to a destination table.`

### Issue description

```
>> dbt archive --target prod
Running with dbt=0.13.0
WARNING: Configuration paths exist in your dbt_project.yml file which do not apply to any resources.
Found 61 models, 844 tests, 1 archives, 0 analyses, 111 macros, 0 operations, 10 seed files, 0 sources

13:56:09 | Concurrency: 8 threads (target='prod')
13:56:09 |
13:56:09 | 1 of 1 START archive project_name.dataset_source_name.table_name --> project_name.dataset_archive_name.table_name [RUN]
13:57:07 | 1 of 1 ERROR archiving dataset_source_name.table_name --> dataset_archive_name.table_name [ERROR in 57.37s]
13:57:07 |
13:57:07 | Finished running 1 archives in 59.96s.

Completed with 1 errors:

Database Error in archive table_name (dbt_project.yml)
  Response too large to return. Consider setting allowLargeResults to true in your job configuration. For more information, see https://cloud.google.com/bigquery/troubleshooting-errors

Done. PASS=0 ERROR=1 SKIP=0 TOTAL=1
```

### System information
The output of `dbt --version`:
```
installed version: 0.13.0
   latest version: 0.13.0

Up to date!
```

The operating system you're running on: Mac OS X 10.14.4

The python version you're using (probably the output of `python --version`): Python 3.7.1

### Steps to reproduce

DBT archive on a table > ~280 gigs within BigQuery.

### Feature description
DBT archives could also benefit from Partitioning / Clustering in BiqQuery. This will improve cost for querying and archiving in the future.

### Who will this benefit?

DBT archive adoption for bigquery users.

### PR DESCRIPTION
**Pull request by drewbanin**
### Archival + Merge

This PR rounds out the changes for the 0.14.0 implementation of Archival.

**Archive changes**:
- Use a `merge` statement where supported (closes #1339)
- Build the initial archive table with a `create table as ( ... )` statement
- Use supplied timestamp column type as `valid_to` type in archive table (closes #938)
- Make it possible to define user-supplied archive strategies
- Set the `ParsedNode` `database` and `schema` fields correctly
- 

**Other changes**:
- Set a destination table when generating BigQuery temp tables (closes #1423)
- Make `expand_target_column_types` take a relation instead of a string
- Add timing info to `pytest` config

**Breaking changes**:
The `expand_target_column_types` adapter method now takes a relation, not a string. This is also used in the dbt-utils implementation of [insert_by_period](https://github.com/fishtown-analytics/dbt-utils/blob/master/macros/materializations/insert_by_period_materialization.sql#L134)

### CODE REVIEW

File: test/integration/004_simple_archive_test/test-archives-bq/archive.sql [hunk 1]
```diff
@@ -9,6 +9,6 @@
             updated_at='updated_at',
         )
     }}
-    select * from `{{database}}`.`{{schema}}`.seed
+    select * from `{{target.database}}`.`{{schema}}`.seed
```
Comment (beckjake) [inline test/integration/004_simple_archive_test/test-archives-bq/archive.sql]: why `target.database` and not `database`? Or alternatively, why `schema` and not `target.schema`?

File: test/integration/032_concurrent_transaction_test/test_concurrent_transaction.py [hunk 2]
```diff
@@ -40,6 +41,8 @@ def run_select_and_check(self, rel, sql):
                 self.query_state[rel] = 'bad'
 
         except Exception as e:
+            logger.info("Caught exception: {}".format(e))
```
Comment (beckjake) [inline test/integration/032_concurrent_transaction_test/test_concurrent_transaction.py]: This is debug stuff, right? Do you want to remove it?

File: plugins/postgres/dbt/include/postgres/macros/adapters.sql [hunk 3]
```diff
@@ -91,3 +91,19 @@
 {% macro postgres__current_timestamp() -%}
   now()
 {%- endmacro %}
+
+{% macro postgres__archive_get_time() -%}
+  {{ current_timestamp() }}::timestamp without time zone
+{%- endmacro %}
+
+{% macro postgres__make_temp_relation(base_relation, suffix) %}
```
Comment (drewbanin) [inline plugins/postgres/dbt/include/postgres/macros/adapters.sql]: @beckjake check this out - I override the `make_temp_relation` macro for pg/redshift to just return a Relation with an identifier. This means that we can use this temp relation to:
1. Build a temp table using `create_table_as(True, tmp_relation)`
2. Query the relation directly
3. Pass it around to adapter methods

File: core/dbt/adapters/base/impl.py [hunk 4]
```diff
@@ -584,22 +585,22 @@ def valid_archive_target(self, relation):
             dbt.exceptions.raise_compiler_error(msg)
 
     @available.parse_none
-    def expand_target_column_types(self, temp_table, to_relation):
+    def expand_target_column_types(self, from_relation, to_relation):
```
[the remark this answers is not in the document]
Comment (drewbanin) [reply core/dbt/adapters/base/impl.py]: do you think we should instead
1. deprecate this method
2. use `expand_column_types` directly henceforth?

this method was just a weirdo adapter method that converts a string into a Relation, then calls `expand_column_types`

[the remark this answers is not in the document]
Comment (beckjake) [reply core/dbt/adapters/base/impl.py]: Nah, I think this is fine to just document for that reason.

File: test/integration/004_simple_archive_test/test-archives-bq/archive.sql (hunk 1 above)
Comment (drewbanin) [reply test/integration/004_simple_archive_test/test-archives-bq/archive.sql]: I changed the archive parsing code to set `database` and `schema` from the `target_database` and `target_schema` configs. This meant that `{{database}}` would interpolate to the target database for the archive, which can be different than the database where the seed lives. Working on a test that exercises this, will remove the extraneous changes though - most of the `target.database` additions i made here aren't actually necessary

**Pull request merged**